lib/Basics/Guarded.h

```diff
@@ -0,0 +1,276 @@
+ */
+
+namespace arangodb {
+
+class Mutex;
+
+template <class T, class L>
+class MutexGuard {
+ public:
+  explicit MutexGuard(T& value, L mutexLock);
+  ~MutexGuard() = default;
+
+  MutexGuard(MutexGuard&&) noexcept = default;
+  auto operator=(MutexGuard&&) noexcept -> MutexGuard& = default;
+
+  auto get() const noexcept -> T&;
+  auto operator->() const noexcept -> T*;
+
+ private:
+  struct nop {
+    void operator()(T*) {}
+  };
+  std::unique_ptr<T, nop> _value;
+  L _mutexLock;
+};
+
+template <class T, class L>
+MutexGuard<T, L>::MutexGuard(T& value, L mutexLock)
+    : _value(&value), _mutexLock(std::move(mutexLock)) {
+  if (ADB_UNLIKELY(!_mutexLock.owns_lock())) {
+    throw std::invalid_argument("Lock not owned");
+  }
+}
+
+template <class T, class L>
+auto MutexGuard<T, L>::get() const noexcept -> T& {
+  return *_value;
+}
+
+template <class T, class L>
+auto MutexGuard<T, L>::operator->() const noexcept -> T* {
+  return std::addressof(get());
+}
+
+template <class T, class M = std::mutex, template <class> class L = std::unique_lock>
+class Guarded {
+ public:
+  using value_type = T;
+  using mutex_type = M;
+  using lock_type = L<M>;
+
+  explicit Guarded(T&& value = T());
+  template <typename... Args>
+  explicit Guarded(Args&&...);
+
+  ~Guarded() = default;
+  Guarded(Guarded const&) = delete;
+  Guarded(Guarded&&) = delete;
+  auto operator=(Guarded const&) -> Guarded& = delete;
+  auto operator=(Guarded&&) -> Guarded& = delete;
+
+  template <class F, class R = std::invoke_result_t<F, value_type&>>
+  auto doUnderLock(F&& callback) -> R;
+  template <class F, class R = std::invoke_result_t<F, value_type&>>
+  auto doUnderLock(F&& callback) const -> R;
+
+  template <class F, class R = std::invoke_result_t<F, value_type&>,
+            class Q = std::conditional_t<std::is_void_v<R>, std::monostate, R>>
+  [[nodiscard]] auto tryUnderLock(F&& callback) -> std::optional<Q>;
+  template <class F, class R = std::invoke_result_t<F, value_type&>,
+            class Q = std::conditional_t<std::is_void_v<R>, std::monostate, R>>
+  [[nodiscard]] auto tryUnderLock(F&& callback) const -> std::optional<Q>;
+
+  // get a copy of the value, made under the lock.
+  template <typename U = T, std::enable_if_t<std::is_copy_constructible_v<U>, int> = 0>
+  auto copy() const -> T;
+
+  // assign a new value using operator=, under the lock.
+  template <class U, std::enable_if_t<std::is_assignable_v<T, U>, int> = 0>
+  void assign(U&&);
+
+  using mutex_guard_type = MutexGuard<value_type, lock_type>;
+  using const_mutex_guard_type = MutexGuard<value_type const, lock_type>;
+
+  auto getLockedGuard() -> mutex_guard_type;
+  auto getLockedGuard() const -> const_mutex_guard_type;
+
+  auto tryLockedGuard() -> std::optional<MutexGuard<value_type, lock_type>>;
+  auto tryLockedGuard() const -> std::optional<MutexGuard<value_type const, lock_type>>;
+  // TODO add more types of "get guard" (like try or eventual)
+
```

Comment:
> the compiler calls operator-> again on the result (see https://timsong-cpp.github.io/cppwp/n4659/over.ref)

Interesting, I was not aware of that. Thanks!

Good idea. This is very convenient, which in turn can increase acceptance. Though I am wondering whether that again opens up a class of accidental errors, e.g.
```
guarded->foo = 10;
guarded->bar = "ten";
```
, where the mutex unlocked in between that would be avoided otherwise. I don't have a feeling yet for how probable such a mistake could be. The scenario of "simply" adding an additional assignment to a single existing one comes to my mind.

I do not have a final opinion on this yet. WDYT?

Reply:
Yes, after writing this I also realized that this is a likely misuse. If you only want to perform a single function call or assignment the example can still be simplified to this onliner `guarded.getLockedGuard()->value = 13` which might be simple enough.

With the additional call to `getLockedGuard()` I would expect it to be very clear that a this usage is definitely suboptimal:
```
guarded.getLockedGuard()->foo = 10;
guarded.getLockedGuard()->bar = "ten";
```
So perhaps it is better to stick with your original design.